File: mobile/lib/domain/community/stimme_tags.dart
/// Zustand an der Stimme — Fahrer-Wahrnehmung, kein Wetter-Orakel.
const kStimmeTagWires = <String>[
  'nass',
  'zu',
  'viel_los',
  'top',
  'baustelle',
];

const kStimmeTagsMax = 3;

Set<String> stimmeTagAllowlist() => {...kStimmeTagWires};

/// Unbekannte Werte fallen weg. Max [kStimmeTagsMax], Reihenfolge bleibt.
List<String> parseStimmeTags(Object? raw) {
  if (raw is! List) return const [];
  final allow = stimmeTagAllowlist();
  final out = <String>[];
  final seen = <String>{};
  for (final e in raw) {
    final t = '$e'.trim().toLowerCase().replaceAll(' ', '_');
    if (!allow.contains(t) || !seen.add(t)) continue;
    out.add(t);
    if (out.length >= kStimmeTagsMax) break;
  }
  return out;
}

List<String> toggleStimmeTag(List<String> current, String wire) {
  final allow = stimmeTagAllowlist();
  if (!allow.contains(wire)) return parseStimmeTags(current);
  final next = [...parseStimmeTags(current)];
  if (next.contains(wire)) {
    next.remove(wire);
    return next;
  }
  if (next.length >= kStimmeTagsMax) return next;
  return [...next, wire];
}

File: src/lib/community/stimmeTags.ts
/** Zustand-Tags an einer Stimme. Unbekannte Werte fallen weg. */

/** Rider condition pills — wet / closed / busy. top/works are not conditions. */
export const STIMME_FORM_TAG_WIRES = ["nass", "zu", "viel_los"] as const;

/** Stored wires still parse historical top/works so old Stimmen stay readable. */
export const STIMME_TAG_WIRES = [
  ...STIMME_FORM_TAG_WIRES,
  "top",
  "baustelle",
] as const;

export type StimmeTagWire = (typeof STIMME_TAG_WIRES)[number];

const ALLOW = new Set<string>(STIMME_TAG_WIRES);
const MAX = 3;

export function parseStimmeTags(raw: unknown): StimmeTagWire[] {
  if (!Array.isArray(raw)) return [];
  const out: StimmeTagWire[] = [];
  const seen = new Set<string>();
  for (const e of raw) {
    const t = String(e ?? "")
      .trim()
      .toLowerCase()
      .replace(/\s+/g, "_");
    if (!ALLOW.has(t) || seen.has(t)) continue;
    seen.add(t);
    out.push(t as StimmeTagWire);
    if (out.length >= MAX) break;
  }
  return out;
}
